### DISCUSSION TITLE
Can I use import in scripts so I can reuse code in my scripts?

### DISCUSSION THREAD

**Discussion by narath** [freeplane/freeplane#684] [Help] [answered]
FreePlane 1.10.4

Hi everyone, love the new Freeplane website, and so appreciate the discussions here. I've been working on some scripts so I can use Freeplane as my GTD tasking system. It's evolved into a collection of a number of closely related scripts with some code duplication. I'd like to refactor it so I can have a lib/ folder and put some utility methods there, but I've been unable to figure out how to import it correctly.

Here's an example from my scripts folder:

```groovy
// Lessons learned in trying to import another class
// works but not sure how to call the class
import lib.tasking.*

// does not work
// import static lib.tasking.getNextSibling as getNextSibling

// none of these work
def nextNode = getNextSibling(node)
def nextNode = tasking.getNextSibling(node)
ui.informationMessage("Got node: " + nextNode.text)
```

Here's the lib file I have in my `scripts/lib` folder

```groovy
// lib/tasking.groovy
package lib

class tasking {
  static def getNextSibling(node) {
    def nodeNext
    def nodeSiblings = node.parent.getChildren()
    def myPosition = node.parent.getChildPosition(node)
    if (myPosition == (nodeSiblings.size()-1)) {
      nodeNext = node.parent
    } else {
      nodeNext = nodeSiblings.get(myPosition+1)
    }
    return nodeNext
  }
}
```

Thanks in advance,
Narath.

**Comment by macmarrum** [comment, marked as answer]
Hi Narath

In Freeplane, scripts and libraries live in two separate sub-directories of **User Directory**.
_I open **User Directory** from Freeplane: `Tools->Open user directory`_

- freeplane/1.10.x/
  - scripts/
  - lib/

**Example**
I create a file `<user-dir>/lib/Abc.groovy`
```groovy
package io.github.macmarrum

class Abc {
    static test() {
        return 'TEST'
    }
}
```

In a Freeplane script or formula I enter
```groovy
import io.github.macmarrum.Abc

Abc.test()
```

Alternatively
```groovy
import static io.github.macmarrum.Abc.test

test()
```
___
Regarding your `tasking.groovy`, once you make sure it's in `<user-dir>/lib/`, you can change `import lib.tasking.*` to `import lib.tasking`

Also, `import static lib.tasking.getNextSibling` works for me.

___

You might also be interested in
- https://github.com/freeplane/freeplane/discussions/589#discussioncomment-3472452

Regards
Mac

**Comment by narath** [comment]
Thanks so much Mac - really helpful!
